### DISCUSSION TITLE
Plugin background Task

### DISCUSSION THREAD

**Discussion by esatterwhite** [Kong/kong#11287] [Ideas and feature requests]
Maybe this exists, but is there a way for a plugin to run a task in the background on an interval, or is the current solution nginx timers + locks?

**Comment by ADD-SP** [comment]
Yes, the OpenResty timer can be used for background tasks or periodic background tasks.

If you want to avoid two periodic background tasks to be overlapped, you can use
```
kong.timer.named_every("task-name", callback_func, interval)
```
`kong.timer` is the instance of [Kong/lua-resty-timer-ng](https://github.com/Kong/lua-resty-timer-ng), it can create a recurrent timer, and it never will be overlapped. For more API, please see the docs of lua-resty-timer-ng.

**lua-resty-timer-ng was added in Kong Gateway v3.**

**Comment by esatterwhite** [comment]
Is this callback function executed on the main lua thread? If this were a long running task would it block incoming requests from being processed?

**Reply by ADD-SP** [reply]
No, it doesn't block the main lua thread, it runs on the other coroutine (not a OS thread).
